### ISSUE TITLE
Upgrade to Saxon 12

### ISSUE THREAD

**Issue by ebruchez** [orbeon/orbeon-forms#976]
We are now (Orbeon Forms 2017.1 included) using Saxon 9.1.

~~Saxon 9.7 was released on November 27, 2015~~
~~Saxon 9.8.0.2 was released on June 8, 2017~~
~~Saxon 9.9.0.2 was released on November 7, 2018~~
~~Saxon 9.9.1.7 was released on 4 March 2020~~
~~Saxon 12.5 was released on 2 July 2024~~

See the [latest Saxon release](http://www.saxonica.com/products/latest.xml).

Since Saxon 9.2, the following features at least have been removed from the open source version:

1. Saxon extension functions (`saxon:*`)
2. EXSLT extension functions
3. Extensibility using reflexion (calls to Java)
4. dom4j support (this is ours so no problem)

We seem to use often:

- `saxon:parse()` (51)
- `saxon:evaluate()` (21)
- `saxon:serialize()` (38)

And the following more rarely:
- `saxon:path()` (1 in FB, not an issue as we can use our own function)
- `saxon:sort()` (1 in FB, XPath 3 has a [`sort(` function](https://www.w3.org/TR/xpath-functions-31/#func-sort))
- ~~`saxon:item-at()` (2 in FB)~~
- `saxon:expression()` (2 in FB, 1 in sandbox)
- `saxon:string-to-base64Binary()` (2 in FR, 1 in sample)
- `saxon:eval()` (1 in sandbox)
- `saxon:compile-stylesheet()` (1 in sample)
- `saxon:in-summer-time()` (3 in tests)

We don't seem to be using EXSLT at all.

What we need to do:

- Restore the missing `saxon:*` functions we really need if any (in a different namespace).
- Retrofit extensibility using reflexion (calls to Java). `JavaExtensionLibrary` seems to be gone since 9.2.
    - see also #2214
- Adapt our code to Saxon 9.8 APIs (documented or not). In particular, note that Michael Kay mentions a [new Sequence class to model all XPath values, replacing the old clumsy Value/ValueRepresentation split](http://saxon-xslt-and-xquery-processor.13853.n7.nabble.com/Saxon-9-5-is-available-td11705.html).

It seems feasible.

**Comment by ebruchez** [comment]
Task: when upgrading, also check uses of `DecimalValue.makeDecimalValue` which since Saxon 9.2 can throw exceptions. For validation, we would like a version of this which doesn't throw.

**Comment by ebruchez** [comment]
Also noting that starting with Saxon 6, Saxon HE includes "basic" XPath 3/3.1 support. This means:

> all XPath 3.1 features other than schema-awareness and higher-order functions. This includes an implementation of maps and arrays, and support for JSON.

It's unclear whether higher-order functions are entirely out of the open source code base, or if they could be brought back.

For reference, this is what XPath 3.0 brings in:

- String concatenation operator ||
- Simple mapping operator !
- Casting is allowed from a string to a union or list type
- Union types, provided they meet certain rules, can be used as a SequenceType
- Dynamic function call (functions are first class values (items) in the data model)
- Function literals, for example substring#2
- Inline functions, for example function($i) {$i*$i}
- Partial application of functions, for example concat('$', ?)
- Let expressions, for example let $v := expr return f($v)
- EQNames (Q{uri}local) wherever QNames are allowed

And XPath 3.1:

- Support for maps and arrays
- Functions to read and write data in JSON format
- A new operator (=>) to allow function chaining
- Various new functions including random-number-generator(), sort(), parse-ietf-date().

**Comment by ebruchez** [comment]
Also pay attention to #3115 when upgrading.

**Comment by ebruchez** [comment]
Noting that `PathMap` is also absent, but see #3296.

**Comment by ewg118** [comment]
Just chiming in to say I'm interested in this. 9.8 HE now does XSLT 3.0 functions like parsing JSON

**Comment by avernet** [comment]
Ethan, I've not sure if you've used the Orbeon Forms build-in JSON functions already, but you might find [those functions](https://doc.orbeon.com/xforms/xpath/extension-json.html) useful.

**Comment by ewg118** [comment]
Are these accessible within XSLT transformations? I have not used them, but I have used the xforms:submission processor to get JSON from a web service and passed the resulting XML response model into XSLT for transformation

**Comment by ebruchez** [comment]
Yes they are accessible from XSLT in the `http://www.orbeon.com/oxf/pipeline` namespace.

Note that these functions follow the XForms format and so convert JSON to XML differently from XPath 3.

**Comment by ewg118** [comment]
Thanks. I'll check it out. Apropos, https://twitter.com/adamretter/status/903604656688988162 : probably time to have a W3C discussion on standardizing JSON->XML serialization

**Comment by ebruchez** [comment]
Possibly. The thing is that the different serializations proposed so far have different purposes. In XForms, the idea is to serialize to a format which lends itself to fairly natural XPath expressions on the data. XSLT and the IBM serialization went in another direction.

**Comment by ebruchez** [comment]
Also if/when we do this, it would probably be time to split XSLT from XPath.

We could even migrate XSLT usages of Saxon separately from migrating XPath usages.

We would have a single version of the XPath processor which should ultimately run on the client and the server. How that is done is TBD. Saxon now uses Java 8, and it's unclear whether using Scalagen, which is old an unmaintained, is realistic in this case. It worked great when experimenting for [Darius XPath](http://ebruchez.github.io/darius-xpath.js/) a few years ago!

**Comment by ebruchez** [comment]
"Saxon 9.9.1.7, released on 4 March 2020"

**Comment by ebruchez** [comment]
With Saxon 10, higher-order functions are available in Saxon HE.

**Comment by ebruchez** [comment]
As part of  #1221,  we are also upgrading to Saxon 10's XPath processor, which we will use both on the client and the server.

What Saxon APIs do we use? For example, from :

- `ControlXPathSupport`: `XPathCache.evaluateAsAvt` and `XPathCache.evaluateAsStringOpt`
- `XPathCache.getXPathExpression`
- `PooledXPathExpression` takes a Saxon `org.orbeon.saxon.sxpath.XPathExpression`
- `org.orbeon.oxf.util.XPath.compileExpression`
    - also used by `StaticBind`, `PathMapXPathAnalysis`

So `XPath` is the big interface point.

Things to adapt:

- `ValueRepresentation` seems gone from Saxon 10 (use `Sequence`?)
- TODO

**Comment by ebruchez** [comment]
As a start we are conservatively moving to Saxon 10 for the Scala.js target.

The following needs to be changed due to changes in Saxon packages, classes, and methods:

- [  ] `SimplePath`
- [  ] `SaxonUtils`
- [  ] `Implicits`
- [  ] `XPath`
- TODO

**Comment by ebruchez** [comment]
As we plan to fully move to Saxon 10 soon as well, we work in two phases:

1. Duplicate files form the JVM to the JS target, and adjust the JS (Saxon 10) versions.
2. Once we are ready to move the JVM target to Saxon 10, we can move the JS versions to `shared`.

**Comment by ebruchez** [comment]
- [ ] `PathMap`
     - Q: How do we do this? Do we still keep Saxon 9 around and our old code for a while? But if so, we need Saxon 10 in a different package, both JVM and Scala.js.

**Comment by ebruchez** [comment]
We have upgraded to Saxon 10 for the JavaScript runtime, but not the JVM runtime. This won't be done for 2021.1.

**Comment by ebruchez** [comment]
- [ ] we could already migrate our XPath function libraries to using `XPathFunctionAnnotationMacro`, even while still targeting Saxon 9
    - the macro might need to generate something different for Saxon 9
    - we can detect this in the macro or we can have another macro for that use case
    - this step is necessary anyway as part of the move to Saxon 10

**Comment by ebruchez** [comment]
[customer](https://3.basecamp.com/3600924/buckets/7545500/messages/5198582962) using `saxon:compile-stylesheet()` and `saxon:transform()`.

**Comment by ebruchez** [comment]
There is hope for Scala 3 [Macro Annotations](https://github.com/lampepfl/dotty/pull/16392).
